File: p12.cpp
#include <iostream>
#include <cmath>
#include <stdint.h>

using std::cout;
using std::endl;

uint32_t numberOfDivisors(int number) {
    if (number == 0) return 0;
    uint64_t nod = 0;
    uint64_t sr = (uint64_t)sqrt(number);
    for (uint64_t i=1; i <= sr; i++) {
        if (number % i == 0) {
            nod += 2;
        }
    }
    if (sr * sr == number) nod--;
    return nod;
}

int main() {
    uint64_t number = 0;
    uint64_t i = 1;
    while (numberOfDivisors(number) < 500) {
        number += i;
        i++;
        //if (i % 1000 == 0) { cout << "triangle number("<<i<<") is " << number << endl; }
    }
    cout << "number = " << number << endl;
    return 0;
}

File: p12.py
#!/usr/bin/python2.6

from primes import get_prime_factors, number_of_divisors, get_n_primes
import math

def num_of_divisors(number):
    nod = 0
    sqrt = int(math.sqrt(number))
    for i in range(1,sqrt):
        if number % i == 0:
            nod += 2
    if (sqrt * sqrt == number):
        nod -= 1
    return nod


if __name__=="__main__":
    number = 0
    i = 1   
    while number_of_divisors(get_prime_factors(number)) < 500:
        #if i % 1000 == 0: print i, number
        number += i
        i +=1

    print number
